Toolchain/CMake/ClangToolchain.txt.in

```diff
@@ -0,0 +1,40 @@
+list(APPEND CMAKE_MODULE_PATH "@SERENITY_SOURCE_DIR@/Toolchain/CMake")
+
+set(CMAKE_SYSTEM_NAME SerenityOS)
+set(CMAKE_SYSTEM_PROCESSOR "@SERENITY_ARCH@")
+set(CMAKE_CROSS_COMPILING 1)
+
+set(SERENITYOS 1)
+set(triple @SERENITY_ARCH@-pc-serenity)
+set(TOOLCHAIN_ROOT @SERENITY_SOURCE_DIR@/Toolchain/Local/clang/@SERENITY_ARCH@)
+set(TOOLCHAIN_PATH ${TOOLCHAIN_ROOT}/bin)
+
+# where to read from/write to
+set(CMAKE_SYSROOT @SERENITY_BUILD_DIR@/Root)
+set(CMAKE_STAGING_PREFIX @SERENITY_BUILD_DIR@/Root/usr/local)
+set(CMAKE_INSTALL_PREFIX @SERENITY_BUILD_DIR@/Root/usr/local)
+set(CMAKE_INSTALL_DATAROOTDIR @SERENITY_BUILD_DIR@/Root/usr/local/share)
+
+set(CMAKE_C_COMPILER ${TOOLCHAIN_PATH}/clang)
+set(CMAKE_C_COMPILER_TARGET ${triple})
+set(CMAKE_CXX_COMPILER ${TOOLCHAIN_PATH}/clang++)
+set(CMAKE_CXX_COMPILER_TARGET ${triple})
+set(CMAKE_ASM_COMPILER ${TOOLCHAIN_PATH}/clang)
+set(CMAKE_ASM_COMPILER_TARGET ${triple})
+set(CMAKE_LINKER ${TOOLCHAIN_PATH}/ld.lld)
+set(CMAKE_RANLIB ${TOOLCHAIN_PATH}/llvm-ranlib)
+set(CMAKE_STRIP ${TOOLCHAIN_PATH}/llvm-strip)
+set(CMAKE_AR ${TOOLCHAIN_PATH}/llvm-ar)
+set(CMAKE_CXXFILT ${TOOLCHAIN_PATH}/llvm-cxxfilt)
+# FIXME: Persuade LLVM maintainers to add `--update-section` to llvm-objcopy, as it's required for the kernel symbol map.
+set(CMAKE_OBJCOPY ${TOOLCHAIN_ROOT}/binutils/bin/@SERENITY_ARCH@-pc-serenity-objcopy)
+
+set(CMAKE_EXE_LINKER_FLAGS_INIT "-Wl,--hash-style=gnu,-z,relro,-z,now,-z,noexecstack,-z,max-page-size=0x1000")
+
+set(CMAKE_C_COMPILER_WORKS YES)
+set(CMAKE_CXX_COMPILER_WORKS YES)
```

Comment:
Why is this needed? This shouldn't generally be messed with.

Reply:
We use the Toolchain file to build things required to link a real executable, such as startup files (Userland/Libraries/LibC/crt0_shared.cpp for example).  So from a clean build it won't work. I was thinking of replacing this with the flag that says "use a static lib to check if the compiler works".

Reply:
These variables come from [here](https://github.com/Kitware/CMake/blob/master/Modules/CMakeTestCCompiler.cmake#L38) and [here](https://github.com/Kitware/CMake/blob/master/Modules/CMakeTestCXXCompiler.cmake#L38). Can the toolchain really not compile those simple test sources?

Reply:
```
CMake Error at /usr/share/cmake-3.17/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "/home/andrew/serenity/Toolchain/Local/clang/i686/bin/clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/andrew/serenity/Build/i686/CMakeFiles/CMakeTmp

    Run Build Command(s):/usr/bin/ninja cmTC_2f4eb && [1/2] Building C object CMakeFiles/cmTC_2f4eb.dir/testCCompiler.c.o
    [2/2] Linking C executable cmTC_2f4eb
    FAILED: cmTC_2f4eb
    : && /home/andrew/serenity/Toolchain/Local/clang/i686/bin/clang --target=i686-pc-serenity --sysroot=/home/andrew/serenity/Build/i686/Root   -Wl,--hash-style=gnu,-z,relro,-z,now,-z,noexecstack,-z,max-page-size=0x1000 CMakeFiles/cmTC_2f4eb.dir/testCCompiler.c.o  -o cmTC_2f4eb   && :
    ld.lld: error: cannot open crt0.o: No such file or directory
    ld.lld: error: cannot open crti.o: No such file or directory
    ld.lld: error: unable to find library -lc
    ld.lld: error: cannot open crtn.o: No such file or directory
    clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
    ninja: build stopped: subcommand failed.
```
Nope it cannot. LibC comes from the build, crt0.o comes from the build, as does crti.o and crtn.o.

We provide crti.o and crtn.o from the asm sources [here](https://github.com/SerenityOS/serenity/tree/fa4eeca0ac3330ba6fee226b315ca06652a7ea11/Userland/Libraries/LibC/arch/i386), while crt0.o comes from [this cpp file](https://github.com/SerenityOS/serenity/blob/fa4eeca0ac3330ba6fee226b315ca06652a7ea11/Userland/Libraries/LibC/crt0.cpp)

So we really don't have the things necessary to run those try compile checks before building LibC. 

So, going to probably replace this with [CMAKE_TRY_COMPILE_TARGET_TYPE](https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html#cmake-try-compile-target-type) set to STATIC.

Reply:
Commit that introduces CMAKE_TRY_COMPILE_TARGET_TYPE. https://github.com/Kitware/CMake/commit/7f1bd9fe6910f7633d98dec018cc01331a46b87e

> This will be useful for cross-compiling toolchains that cannot
> link a binary without custom flags or scripts.

I don't believe that's true for SerenityOS, this feels like it'd be a hack. These object files could be built ahead of time, before the OS child build tree is configured.  
If I'm reading your reply correctly, it sounds like LibC should be moved into its own build as well.

Reply:
I'm not sure I'm seeing the value in creating yet another toolchain file and build stage right now just for the C Library build. That build would definitely need that TRY_COMPILE_TARGET_TYPE setting, because it's building the C library.

I can leave a FIXME to take a look at that. There's definitely some things that LibC/CMakeLists.txt does that are questionable.. like creating explicit static and shared libraries, etc. We include separate object libs for like. `ubsan` and `ssp` for `__stack_chk_fail` and some ... interesting add_dependencies there that make it a bit of a ball of wax on its own.

Reply:
I see. Aiming for the cleanest possible solution would be more effort than worth right now. Using the `STATIC_LIBRARY` option with a FIXME is a good compromise for the sake of incremental improvement.

Reply:
For some historical context, we used to actually build the C library before libstdc++ as part of the toolchain build, because it was observed that libstdc++ actually needed a working C library in order to compile/link. But, turns out that it only needed the headers after Ben looked into it some more.

https://github.com/SerenityOS/serenity/commit/5f724b6ca1aae3a5a8c7189069649e8a9347cca2#diff-27e8b38fa0c632f71ab59fc31a34847ece5cbbc670f74514d313217fb25637a3L244-R252

But supporting that pre-libstdc++ build caused quite a few headaches, as things like `std::initailizer_list`, or `__cxa_demangle` were not available. Since that time we've moved to CMake, so the build tool logic to pre-build LibC seems to have been lost to time (and like > 10k commits).